pygpu/ufuncs.py

```diff
@@ -0,0 +1,1145 @@
+
+    # TODO: can CPU memory handed directly to kernels?
+    if not isinstance(a, GpuArray):
+        a = numpy.asarray(a)
+        if a.flags.f_contiguous and not a.flags.c_contiguous:
+            order = 'F'
+        else:
+            order = 'C'
+        a = array(a, dtype=a.dtype, copy=False, order=order, context=ctx,
+                  cls=cls)
+
+    if a.dtype == numpy.dtype('float16'):
+        # Gives wrong results currently, see
+        # https://github.com/Theano/libgpuarray/issues/316
+        raise NotImplementedError('float16 currently broken')
+
+    prom_dtypes_in, result_dtypes = ufunc_dtypes(ufunc_name, [a.dtype])
+    prom_dtype_in = prom_dtypes_in[0]
+    result_dtype1, result_dtype2 = result_dtypes
+
+    # This is the "fallback signature" case, for us it signals failure.
+    # TypeError is what Numpy raises, too, which is kind of wrong
+    if prom_dtype_in == numpy.dtype(object):
+        raise TypeError('input dtype {!r} invalid for ufunc {!r}'
+                        ''.format(a.dtype.name, ufunc_name))
+
+    # Convert input such that the kernel runs
+    # TODO: can this be avoided?
+    a = a.astype(prom_dtype_in, copy=False)
+
+    if out1 is None:
+        out1 = empty(a.shape, dtype=result_dtype1, context=ctx, cls=cls)
+    else:
+        # TODO: allow larger dtype
+        if out1.dtype != result_dtype1:
+            raise ValueError('`out1.dtype` != result dtype: {!r} != {!r}'
+                             ''.format(out1.dtype.name, result_dtype1.name))
+    if out2 is None:
+        out2 = empty(a.shape, dtype=result_dtype2, context=ctx, cls=cls)
+    else:
+        if out2.dtype != result_dtype2:
+            raise ValueError('`out2.dtype` != result dtype: {!r} != {!r}'
+                             ''.format(out2.dtype.name, result_dtype2.name))
+
+    # C result dtypes for casting
+    c_res_dtype1 = dtype_to_ctype(result_dtype1)
+
+    oper = ''
+
+    if ufunc_name == 'frexp':
+        oper = 'res = ({rdt1}) frexp(a, &out2)'.format(rdt1=c_res_dtype1)
+    elif ufunc_name == 'modf':
+        oper = 'res = ({rdt1}) modf(a, &out2)'.format(rdt1=c_res_dtype1)
+
+    if not oper:
+        raise ValueError('`ufunc_name` {!r} does not represent a unary ufunc'
+                         ''.format(ufunc_name))
+
+    a_arg = as_argument(a, 'a', read=True)
+    args = [arg('res', out1.dtype, write=True),
+            arg('out2', out2.dtype, write=True),
+            a_arg]
+
+    ker = GpuElemwise(ctx, oper, args, convert_f16=True)
+    ker(out1, out2, a)
+    return out1, out2
+
+
+MISSING_UFUNCS = [
+    'conjugate',  # no complex dtype yet
+    ]
+
+
+UFUNC_SYNONYMS = [
+    ('absolute', 'abs'),
+    # ('conjugate', 'conj'),
+    ('deg2rad', 'degrees'),
+    ('rad2deg', 'radians'),
+    ('true_divide', 'divide'),
+    ('maximum', 'fmax'),  # TODO: differ in NaN propagation in numpy, doable?
+    ('minimum', 'fmin'),  # TODO: differ in NaN propagation in numpy, doable?
```

Comment:
Adjusting NaN propagation is doable with a bit of C code.  If you don't want to do it, I'd rather we don't provide fmax/fmin than provide them with the wrong comportment.